extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/transaction/interceptor/MongoTransactionalInterceptor.java

```diff
@@ -0,0 +1,136 @@
+package io.quarkus.mongodb.panache.transaction.interceptor;
+
+import javax.annotation.Priority;
+import javax.inject.Inject;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+import javax.transaction.HeuristicMixedException;
+import javax.transaction.HeuristicRollbackException;
+import javax.transaction.RollbackException;
+import javax.transaction.Status;
+import javax.transaction.Synchronization;
+import javax.transaction.SystemException;
+import javax.transaction.TransactionManager;
+import javax.transaction.TransactionSynchronizationRegistry;
+
+import com.mongodb.ReadConcern;
+import com.mongodb.ReadConcernLevel;
+import com.mongodb.ReadPreference;
+import com.mongodb.TransactionOptions;
+import com.mongodb.WriteConcern;
+import com.mongodb.client.ClientSession;
+import com.mongodb.client.MongoClient;
+
+import io.quarkus.mongodb.panache.transaction.MongoTransactionException;
+import io.quarkus.mongodb.panache.transaction.Transactional;
+
+@Interceptor
+@Transactional
+@Priority(Interceptor.Priority.PLATFORM_BEFORE + 200)
+public class MongoTransactionalInterceptor {
+    private static final Object transactionKey = new Object();
+
+    @Inject
+    MongoClient client;
+
+    @Inject
+    TransactionSynchronizationRegistry tsr;
+
+    @Inject
+    TransactionManager transactionManager;
+
+    @AroundInvoke
+    public Object intercept(InvocationContext ic) throws Exception {
+        // if we are in a sub-method call we just proceed with the invocation
+        // note that nested transactions are not supported
+        // maybe we need to rely on clientSession.hasActiveTransaction ?
+        if (isInTransaction()) {
+            return ic.proceed();
+        }
+
+        transactionManager.begin();
+        Transactional transactional = getTransactional(ic);
+        TransactionOptions.Builder txnOptions = TransactionOptions.builder();
+        if (!transactional.readConcern().equals("<default>")) {
+            ReadConcern readConcern = new ReadConcern(ReadConcernLevel.fromString(transactional.readConcern()));
+            txnOptions.readConcern(readConcern);
+        }
+        if (!transactional.writeConcern().equals("<default>")) {
+            WriteConcern writeConcern = WriteConcern.valueOf(transactional.writeConcern());
+            txnOptions.writeConcern(writeConcern);
+        }
+        if (!transactional.readPreference().equals("<default>")) {
+            ReadPreference readPreference = ReadPreference.valueOf(transactional.readPreference());
+            txnOptions.readPreference(readPreference);
+        }
+
+        try (ClientSession clientSession = client.startSession()) {
+            MongoTransaction transaction = getMongoTransaction(clientSession);
+            // TODO put the transaction somewhere ...
+            clientSession.startTransaction(txnOptions.build());
+            ClienSessionHolder.setClientSession(clientSession);
```

Comment:
You don't need that: you can obtain it from the `TransactionManager`

Reply:
I try to obtain the transaction from the transaction manager and it shows that there is an issue somewhere: `java.lang.IllegalStateException: No transaction is running`.
I need to have a closer look at this ...

Reply:
I now get the transaction from the `TransactionSynchrinizationRegistry`.